### PR TITLE
Feature/ecommerce api

### LINKED ISSUE #504: [S3.3]: E-commerce service points
### Link to Epic

Epic: https://github.com/Open-Study-College/osc/issues/437

### Description

n/a

### Linked Tasks
#505 

_No response_

### LINKED ISSUE #505: Create Mappings for E-commerce
### Link task to story

Story: https://github.com/Open-Study-College/osc/issues/504

### Description

Mapping data structure between third party api and OSC api

### LINKED ISSUE #506: [S3.4]: E-commerce enpoint
### Link to Epic

Epic: https://github.com/Open-Study-College/osc/issues/437

### Description

n/a

### Linked Tasks

#507 
#508 
#509 
#510 
#511 

_No response_

### LINKED ISSUE #507: Set-up duplicate repo
### Link task to story

Story: https://github.com/Open-Study-College/osc/issues/506

### Description

n/a

### LINKED ISSUE #508: Import e-commerce library
### Link task to story

Story: https://github.com/Open-Study-College/osc/issues/506

### Description

n/a

### LINKED ISSUE #509: Build schema
### Link task to story

Story: https://github.com/Open-Study-College/osc/issues/506

### Description

n/a

### LINKED ISSUE #510: Build resolvers
### Link task to story

Story: https://github.com/Open-Study-College/osc/issues/506

### Description

n/a

### PR DESCRIPTION
**Pull request by Tom-Pearce**
- Closes #504 
- Closes #505  
- Closes #506 
- Closes #507 
- Closes #508 
- Closes #509 
- Closes #510 

## 📝 Description

First iteration of the Ecommerce API

## ⛳️ Current behavior (updates)

None

## 🚀 New behavior

Connects to Shopify and Algolia, making the API SDKs available via Apollo's context.
Able to query a basic products endpoint via the gateway

## 📝 Additional Information
Also includes a refactor of the gateway to account for the new `OSC-API` utility package.

### CODE REVIEW

**Review by hi-im-si — CHANGES_REQUESTED** [review]
Do we need to keep the shape of the data that is in algolia for the shopify products e.g.
```
{
  "hits": [
    {
      "title": "Edexcel Functional Skills Maths Level 2",
      "updated_at": "2021-11-09T00:02:21.000Z",
      "_tags": [
        "id:7438721679615",
        "indexing_token:513bf011106517fae5bb8ca7daaef89b"
      ],
      "handle": "functional-skills-maths-level-2",
      "vendor": "openstudydev",
      "product_type": "",
      "template_suffix": null,
      "sku": "OSC1373",
      "barcode": null,
      "position": 1,
      "requires_shipping": true,
      "taxable": true,
      "inventory_quantity": 0,
      "option1": "Course Material",
      "option2": "Online",
      "option3": null,
      "id": 7438721679615,
      "tags": [
        "import_2021_11_08_234706",
        "Mathematics & Economics"
      ],
      "option_names": [
        "format",
        "study-method"
      ],
      "variants_count": 4,
      "variants_min_price": 369,
      "variants_max_price": 569,
      "variants_inventory_count": 0,
      "product_image": "https://cdn.shopify.com/s/files/1/0609/0519/3727/products/OSC1332-e1584960693138_7b7080af-f545-4853-b607-a1a345526c83.jpg?v=1636416540",
      "published_at": "2021-11-09T00:14:37.000Z",
      "body_html_safe": null,
      "variant_title": "Course Material / Online",
      "inventory_policy": "deny",
      "inventory_management": "shopify",
      "inventory_management_shopify": 1,
      "inventory_available": false,
      "options": {
        "format": "Course Material",
        "study-method": "Online"
      },
      "price": 369,
      "compare_at_price": 0,
      "price_ratio": 0,
.........
..........
........
}
```
or we just doing a basic initial schema?

File: packages/osc-api-ecommerce/src/types/functions.ts [hunk 1]
```diff
@@ -0,0 +1 @@
+
```
[the remark this answers is not in the document]
Comment (Tom-Pearce) [reply packages/osc-api-ecommerce/src/types/functions.ts:1]: I can remove it but probably end up creating it again as soon as we build any functionality out.

**Comment by Tom-Pearce** [comment]
> Do we need to keep the shape of the data that is in algolia for the shopify products e.g.
> 
> ```
> {
>   "hits": [
>     {
>       "title": "Edexcel Functional Skills Maths Level 2",
>       "updated_at": "2021-11-09T00:02:21.000Z",
>       "_tags": [
>         "id:7438721679615",
>         "indexing_token:513bf011106517fae5bb8ca7daaef89b"
>       ],
>       "handle": "functional-skills-maths-level-2",
>       "vendor": "openstudydev",
>       "product_type": "",
>       "template_suffix": null,
>       "sku": "OSC1373",
>       "barcode": null,
>       "position": 1,
>       "requires_shipping": true,
>       "taxable": true,
>       "inventory_quantity": 0,
>       "option1": "Course Material",
>       "option2": "Online",
>       "option3": null,
>       "id": 7438721679615,
>       "tags": [
>         "import_2021_11_08_234706",
>         "Mathematics & Economics"
>       ],
>       "option_names": [
>         "format",
>         "study-method"
>       ],
>       "variants_count": 4,
>       "variants_min_price": 369,
>       "variants_max_price": 569,
>       "variants_inventory_count": 0,
>       "product_image": "https://cdn.shopify.com/s/files/1/0609/0519/3727/products/OSC1332-e1584960693138_7b7080af-f545-4853-b607-a1a345526c83.jpg?v=1636416540",
>       "published_at": "2021-11-09T00:14:37.000Z",
>       "body_html_safe": null,
>       "variant_title": "Course Material / Online",
>       "inventory_policy": "deny",
>       "inventory_management": "shopify",
>       "inventory_management_shopify": 1,
>       "inventory_available": false,
>       "options": {
>         "format": "Course Material",
>         "study-method": "Online"
>       },
>       "price": 369,
>       "compare_at_price": 0,
>       "price_ratio": 0,
> .........
> ..........
> ........
> }
> ```
> 
> or we just doing a basic initial schema?

I'm only exposing data that's getting used. I.e. if it's not being exposed in our GraphQL schema I'm not adding it to the internal types to keep it leaner. If new fields are added they can be added to both places as and when they're needed. Could probably do something fancy to make the types more dynamic, but that's something to look into in the future I think. Unlike Prisma, you don't get a unique client that fully typed based on your data unfortunately.

**Comment by stevenbister** [comment]
Can't comment too heavily but looks good! Do we need to include any tests?

**Pull request merged**